SLSA v1.0: add Key Principles page (#528)
slsa-framework/slsa: docs/spec/v1.0/principles.md

# Key principles

<div class="subtitle">

> This page lays out the key guiding principles behind SLSA.

Reviewer: I think it might be worth enumerating the key principles at the top of the page and expanding on them further down, something like:
```suggestion
The key guiding principles behind SLSA are:

- Trust systems, verify artifacts
- Place trust in code, not individuals 
- Prefer explicit attestations over inferences from configuration

These principles are explained below.
```
It somewhat duplicates the navigation bar but may provide a smoother reading experience. WDYT?

Author: Personally I find that less readable, and it's also a bit harder to maintain (easy to get out-of-sync).

Thoughts from other reviewers?

Reviewer: I get it. The current layout looks a bit awkward with a single short sentence followed by a large white space, but perhaps we can do something about that later? I'm approving the PR as-is.

Author: Yup, let's try to improve it in a future PR.

</div>

> ## Automatically verify artifacts, manually verify systems

Author: Alternative title: "Trust systems, [automatically] verify artifacts" (With or without "automatically".)

Any thoughts? That's a bit shorter and might better convey the point.

Reviewer: I much prefer the shorter, alternative title. I would go with "Trust systems, verify artifacts".

Reviewer: LGTM!

> One of the main ideas behind SLSA is to manually verify that a few *systems*,

Reviewer: ```suggestion
One of the main ideas behind SLSA is to manually verify a few *systems*,
```

Once we have a set of trusted systems, we can verify the supply chain of
arbitrarily many *artifacts* by ensuring that artifacts were produced by trusted
systems. This allows SLSA to scale to entire ecosystems and organizations
regardless of size.

For example, a security engineer may analyze the architecture and implementation
of a build system to ensure that it meets the SLSA build requirements. Once the
analysis is complete, the public keys used by the build system to sign
provenance can then be "trusted" up to a given SLSA level. Downstream systems
can then automatically determine that an artifact meets a given level by
verifying that it has provenance signed by that public key.

### Corollary: Minimize the number of trusted systems

A corollary to the previous principle is to minimize the size of the trusted
computing base. Every system we trust adds attack surface and increases the need
for manual security analysis. Where possible:

-   Concentrate trust in shared infrastructure. For example, instead of each
    team maintaining their own build system, use a shared build system. This way
    hardening work can be shared across all teams.

> 

Reviewer: ```suggestion
```
There's a lot of whitespace between these bullets in the preview:
<img width="801" alt="Screenshot 2022-11-04 at 15 18 43" src="https://user-images.githubusercontent.com/13888612/200011053-f8d61136-a5ee-4409-9a3e-c6b3a53908b5.png">

Author: It's the same as the regular paragraph spacing. I switched to "tight" lists - does that look better to you?

Reviewer: Yes, looks much better. Thanks.

    end-to-end signing to avoid the need to trust intermediate distribution
    systems.

## Place trust in code, not individuals

Another idea behind SLSA is to place our trust in source code rather than
individual maintainers who have write access to package registries. Code is
static and analyzable. People, on the other hand, are prone to mistakes,
credential compromise, and sometimes malicious action. Therefore, SLSA aims to
securely trace all software back to source code, without the possibility for an
individual maintainer to tamper with it after it has been committed.

## Prefer explicit attestations over inferences from configuration

SLSA places a strong emphasis on explicit attestations about security
properties, particularly an artifact's provenance, rather than inferring
security properties from how a system is configured.

Theoretically one could configure access control such that the only path from
source to release is through the official channels: the CI/CD system only pulls
from the proper source, package registry only allows access to the CI/CD system,
and so on. In practice, that is almost impossible to get right and keep right.

> There are often over-provisioning, or confused deputy problems, or mistakes.

Reviewer: ```suggestion
There are often over-provisioning, confused deputy problems, or mistakes.
```
?

Even if it is configured properly at one moment, there is no guarantee that it

> will continue to stay so configured. And humans almost always end up getting in

Reviewer: ```suggestion
will continue to stay so. And humans almost always end up getting in
```

the access control lists.

Instead, SLSA prefers explicit attestations about an artifact's provenance.
Access control is still important, but SLSA provides a defense in depth: it is

> not sufficient to be granted access to publish a package, but one must also have

Reviewer: ```suggestion
not sufficient to be granted access to publish a package, one must also have
```
